### ISSUE TITLE
Report cast(or cast to child implementation) when method can be accessed via parent ref

### ISSUE THREAD

**Issue by atulgpt** [detekt/detekt#5999]
## Expected Behavior of the rule
```
// non compliant code
open class A {
    fun doWork() {
        /*no-op*/
    }
}
class B : A()

fun  test(a: Any) {
    (a as? B)?.doWork()
}
```

```
// compliant code
open class A {
    fun doWork() {
        /*no-op*/
    }
}
class B : A()

fun  test(a: Any) {
    (a as? A)?.doWork() //  here we are casting to a base class rather than child implementation
}
```
Or
```
// non compliant code
open class A {
    fun doWork() {
        /*no-op*/
    }
}
class B : A()

fun  test(a: A) {
    (a as B).doWork()
}
```

```
// compliant code
open class A {
    fun doWork() {
        /*no-op*/
    }
}
class B : A()

fun  test(a: A) {
    a.doWork() // here we removed the cast
}
```

## Context
Many times in Android, people cast a view to implementation class like `ImageView`, `Button` etc just to set `setLayoutParams` which can also be set w/o and cast as `setLayoutParams` is available in the base class. Later, when we change the implementation details, the cast breaks at that point.

This will avoid unnecessary cast in some cases and in some places report cast to concrete classes which can help change the implementation w/o changing the cast.

In all the above methods `doWork` method is final

I am not sure though if there is some existing rule or some configuration which can simulate this behaviour

**Comment by marschwar** [comment]
I do not think there is a rule at the moment that would achieve what you are trying to do. While I see where you are coming from, I would argue that this may a little too specific. To me this may be two different rules:

* `UnnecessaryCast` - the instance type implements all methods called in the scope and therefor there is no need to cast to a subtype.
* `CastCouldBeMoreGeneral` (or something along those lines) - Reports a violation when all methods called within the scope of the cast are available at  a parent type or an  interface the type or parent type implements.

Did I understand you correctly? The first one I think could be a good addition. The second one I am much more skeptical of.

**Comment by atulgpt** [comment]
Yes, you are right `UnnecessaryCast` is related to my second example and `CastCouldBeMoreGeneral` is related to my first example. 

> The second one I am much more skeptical of.

Do you think I am missing some case or use-case of this pattern?

**Comment by marschwar** [comment]
When I was trying to come up with a good explanation for this rule, I could not think of a reason why methods on a parent class should be handled any differently than methods of implemented interfaces. 

Also consider a parent class with a very long name. In such case I would probably prefer readability.

But besides that I am not against this rule/these rules even though it may not be as easy to implement.

**Comment by atulgpt** [comment]
> Also consider a parent class with a very long name

I think that will good use-case for `suppression`. The main pain point that I face, is that coupling code with the implementation details makes really hard to change the implementation from one class to another. Especially in Android where cast to a specific view happens at runtime and that crashes the app

**Comment by marschwar** [comment]
Would you like to give it a try?

**Comment by atulgpt** [comment]
Yes I would like to. Looping @BraisGabin and @cortinico for more feedbacks

**Comment by cortinico** [comment]
> I am not sure though if there is some existing rule or some configuration which can simulate this behaviour

We actually do have a really similar rule: `DontDowncastCollectionTypes`
We can probably have a `DontDowncastToImplementationTypes` rule that is similar but more generic and configurable?

**Comment by BraisGabin** [comment]
I like the idea. I think that this idea is very related with could be no null. Should we mix them? A rule to check that the type of the parameters of each function is the correct one. And then the other for casting.

**Comment by atulgpt** [comment]
Not stale

**Comment by atulgpt** [comment]
Hi, @BraisGabin can you explain your comment `I think that this idea is very related with could be no null. Should we mix them? A rule to check that the type of the parameters of each function is the correct one. And then the other for casting.` with examples? Are you referring to the same comment https://github.com/detekt/detekt/issues/5999#issuecomment-1509056048 by @marschwar?

**Comment by BraisGabin** [comment]
TL;DR: I would like to extract some logic from `CanBeNonNullable` to another rule called `WrongParameterType` and don't only check for castings from nullable to non-nullable but for any type of cast. (I'm don't like `WrongParameterType` as a rule name... suggestions are more than welcom)

--- 

I think that the error on this code is not that the cast is not needed

```kotlin
fun test(a: A) {
    (a as B).doWork()
}
```

The problem is that the function doesn't define the correct type. The function should be like this:

```kotlin
fun test(a: B) {
    a.doWork()
}
```

For that reason I related this with `CanBeNonNullable`. That rule flags code like the one below and forces you to change the funcition signature.
```kotlin
fun test(a: A?) {
    a!!.doWork()
}
```

----

Once you have the function like this: 
```kotlin
fun test(a: B) {
    a.doWork()
}
```

We could implement `TypeCouldBeMoreGeneral` but I'm not 100% sure about that one. An Android example:

```kotlin
fun close(b: Button) {
  b.setOnClickListener { activity.finish() }
}
```

It could be changed to:

```kotlin
fun close(v: View) {
  v.setOnClickListener { activity.finish() }
}
```

But even if you *can* do it I'm not sure if you *should* do it.
